Adopt a single allocation API

philsmt opened the pull request:
Alternative implementation to address #5 next to #6 based on Thomas' idea of a single API call for the different allocation scenarios, for direct and per-worker arrays, respectively:

```
MapContext.alloc(shape=None, dtype=None, order=None, fill=None, like=None)
MapContext.alloc_per_worker(shape=None, dtype=None, order=None, fill=None, like=None)
```

It comes in 113 lines (!) lighter than the numpy array API, and I was quite chatty in its docstring 😃

I'm still tempted to fold the second method into a `per_worker` keyword, as it's really _just_ a flag to modify the result. This is nicely highlighted by the fact it's using the same test.
Also, one may make all arguments keyword-only, to ensure nobody is putting any meaning in its order. But it may be a bit too much on pushing a particular style.

philsmt committed:
Add a single API for all direct and per-worker array allocation scenarios

takluyver commented on pasha/context.py:
```diff
@@ -39,56 +39,154 @@ def __init__(self, num_workers):
+                memory order.
+        """
+        if like is not None:
+            shape = shape or like.shape
```
Bit surprising if you pass `shape=0` :wink:

philsmt replied:
Hu, I knew I should've sticked with the recommended `x is None` checks rather than cutting it short 🙄

takluyver replied:
I do like how `arg = arg or default` reads, and there are places where you can get away with it. But yeah, there's a reason the `is None` check is recommended. :slightly_smiling_face:

takluyver commented:
> I'm still tempted to fold the second method into a per_worker keyword, as it's really just a flag to modify the result. This is nicely highlighted by the fact it's using the same test.

That's OK with me if you prefer. I was thinking of them as separate mostly because I wouldn't attach `alloc` to the context,  but you don't seem to be buying that part of my argument. :wink: 

I might even argue that a boolean flag doesn't really add anything over making it explicit:

```python
ctx.alloc_per_worker(100)
ctx.alloc(100, per_worker=True)
ctx.alloc((ctx.n_workers, 100))
```

Though obviously that depends how you use it - e.g. using `alloc(like=x, per_worker=True)` to get an array of shape `(n_workers,) + x.shape` saves a bit of typing.

> Also, one may make all arguments keyword-only, to ensure nobody is putting any meaning in its order. But it may be a bit too much on pushing a particular style.

I might allow shape & dtype as positional args, and make the rest keyword-only - I tend to think that beyond a couple of arguments, it's clearer to pass them by name. But that's not super important.

philsmt committed:
Fold alloc_per_worker into alloc with per_worker argument

philsmt committed:
Make alloc argument except shape/dtype keyword-only

philsmt commented:
> > I'm still tempted to fold the second method into a per_worker keyword, as it's really just a flag to modify the result. This is nicely highlighted by the fact it's using the same test.
> 
> That's OK with me if you prefer. I was thinking of them as separate mostly because I wouldn't attach `alloc` to the context, but you don't seem to be buying that part of my argument. 😉
> 
> I might even argue that a boolean flag doesn't really add anything over making it explicit:

I've found this case, which I find somewhat more expressive with the keyword, as you don't need to switch the method:
```python
one_result = psh.alloc(my_shape, my_dtype, per_worker=True)
other_result = psh.alloc(like=one_result)
```
Both arrays are meant to be `per_worker`, but using `alloc` for the second might be misleading if the names differ.

> I might allow shape & dtype as positional args, and make the rest keyword-only - I tend to think that beyond a couple of arguments, it's clearer to pass them by name. But that's not super important.

Good point, it also gives a minimal amount of compatibility with `numpy`.

This is as far as I'd take it, any more comments? Thanks for the review so far!

D4vidH4mm3r commented:
This interface is pretty neat; `fill` and `like` make sense as keywords. While https://github.com/European-XFEL/pasha/pull/6 may be more familiar to heavy numpy users, I don't think that is worth the extra wrapping paper and API bloat.

I think `per_worker` is fine as a keyword. The docstring does indicate that it is "just" there for conveniently adding an axis without having to recall the number of workers. If it's a separate function, I start wondering if the allocation strategy is somehow different.

philsmt committed:
Move array initialization to low level methods to allow for some optimization

philsmt commented:
Alright, let's keep it like this. Thanks for the reviews!

I've added some final docstring fixes and moved the initialization via `fill` into the low-level method `_alloc`. Quick googling demonstrated a more-than-trivial improvement when allocating large zeroed arrays, see [[1]](https://stackoverflow.com/questions/58076522/why-is-the-speed-difference-between-pythons-numpy-zeros-and-empty-functions-gon) and [[2]](https://vorpus.org/blog/why-does-calloc-exist/). This way we can use `np.zeros` for the heap and actually skip it for shared memory, as `MAP_ANONYMOUS` already guarentees zeroes.

philsmt commented:
Quick testing on a maxwell node gave `7 us` vs `800 ms` for allocating a zeroed `(1000, 1000, 1000)`-shaped array.
